### ISSUE TITLE
using output_layer in seq2seq model?

### ISSUE THREAD

**Issue by sunxiaoben** [tensorflow/tensorflow#11102]
Os：liunx
version: v1.2.0

i am writing a seq2seq model using tensorflow. my problem is 

helper = TrainingHelper(...)
decoder = BasicDecoder(...)
decoder_outputs, final_state, seq_len  = tf.contrib.seq2seq.dynamic_decode(decoder)
rnn_out, sample_ids = decoder_outputs

in traing period, we should calculate the loss:
loss = tf.contrib.seq2seq.sequence_loss(rnn_out, targets, weights)

the rnn_out's shape should be (batch_size, seq_len, target_vocab_size), but (batch_size, seq_len, cell_hidden_size)

how to change the shape ? should i add a output_layer in decoder ? 
if add, how to define a output_layer? i found tf.layers.dense has two params (inputs, units), units may be target_vocab_size, but what about inputs?

In addition, is the output_layer is same as the beamSearcheDecoder when decoding?

**Comment by TrsNium** [comment]
helper = TrainingHelper(...)
decoder = BasicDecoder(...)
decoder_outputs, final_state, seq_len = tf.contrib.seq2seq.dynamic_decode(decoder, output_time_major=True)
rnn_out, sample_ids = decoder_outputs

logits = tf.layers.dense(rnn_out[-1], target_vocab_size)

**Comment by sunxiaoben** [comment]
@TrsNium 
the BasicDecoder has a param output_layer, how can i using this?
decoder = BasicDecoder(..., output_layer=?)

**Comment by TrsNium** [comment]
Sorry, I don't know how to use that  param too.
According to BasicDecoder's code output_layer should be instance of tf.layersLayer.
https://github.com/tensorflow/tensorflow/blob/r1.2/tensorflow/contrib/seq2seq/python/ops/basic_decoder.py

**Comment by paulswtang** [comment]
I think the issue is that `tf.contrib.seq2seq.BasicDecoder` expects an instance of `layers.Dense`, which is an internal class used in `tf.layers.dense`:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/layers/core.py#L41

This can be seen in the tests for `BasicDecoder`:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/seq2seq/python/kernel_tests/basic_decoder_test.py#L38

Note in particular line 55:
` output_layer = layers_core.Dense(output_layer_depth, use_bias=False)`

However, I don't think the Dense layer class is exposed in the Python API as of v1.2. Does anyone have any opinions on how to resolve this?

**Comment by ali01** [comment]
This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. There is also a larger community that reads questions there. Thanks!

**Comment by paulswtang** [comment]
Are there any future plans to change the API of `tf.contrib.seq2seq.BasicDecoder` to accept a callable in the `output_layer` field instead of a `Dense` layer?

It seems kind of unwieldy to restrict the parameter to instances of a class that is inaccessible from the Python API, which is probably what most people are using. (Unless I'm wrong about that, but it doesn't seem to appear in the docs for v1.2.)

**Comment by wcarvalho** [comment]
Any progress made on this? I haven't been able to figure out how to access tensorflow.python.layers.base and thus can't make my own "Dense" class or use theirs? Without that, it seems that the output_layer argument is useless to the general public? Was it never intended to be used by the general public?

**Comment by wcarvalho** [comment]
I got around that pesky error by building Dense as a very basic wrapper around dense:
```
from tensorflow.python.layers import base
class Dense(base.Layer):
    """A wrapper around tf.layers.Dense which lets it be used as a callable class. Refer to https://www.tensorflow.org/versions/master/api_docs/python/tf/layers/dense for more documentation"""
    def __init__(self,
            units,
            activation=None,
            use_bias=True,
            kernel_initializer=None,
            bias_initializer=tf.zeros_initializer(),
            kernel_regularizer=None,
            bias_regularizer=None,
            activity_regularizer=None,
            kernel_constraint=None,
            bias_constraint=None,
            trainable=True,
            name=None,
            reuse=None
        ):
        self.units = units
        self.activation = activation
        self.use_bias = use_bias
        self.kernel_initializer = kernel_initializer
        self.bias_initializer = bias_initializer
        self.kernel_regularizer = kernel_regularizer
        self.bias_regularizer = bias_regularizer
        self.activity_regularizer = activity_regularizer
        self.kernel_constraint = kernel_constraint
        self.bias_constraint = bias_constraint
        self.trainable = trainable
        self.name = name
        self.reuse = reuse
    def call(self, inputs):
        return tf.layers.dense(inputs, units, activation, use_bias, kernel_initializer, bias_initializer, kernel_regularizer, bias_regularizer, activity_regularizer, kernel_constraint, bias_constraint, trainable, name, reuse)

    def __call__(self, x):
        return self.call(x)
```

Hope this helps!

**Comment by WolfNiu** [comment]
I beg to differ since this is an issue caused by unclear documentation. @ali01

**Comment by shawnthu** [comment]
BTW, how can I add multiple dense layers to the output layer, Dense(n) is too simple, I want to add more non-linearity

**Comment by VXU1230** [comment]
> BTW, how can I add multiple dense layers to the output layer, Dense(n) is too simple, I want to add more non-linearity

After you get the decoder rnn output, you can pile several tf.layers.dense() to add non-linearity.
